### PR TITLE
ARROW-480: [Python] Implement RowGroupMetaData.ColumnChunk

### PR DESCRIPTION
**Pull request by Licht-T**
This is the patch for [ARROW-480](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-480).

### CODE REVIEW

File: python/pyarrow/_parquet.pyx [hunk 1]
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
...
+
+    property has_null_count:
+
+        def __get__(self):
+            return bool(self.statistics.has_null_count)
+
+    property has_distinct_count:
+
+        def __get__(self):
+            return bool(self.statistics.has_distinct_count)
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: I didn't think these `bool` casts are needed, but correct me if I'm wrong

File: python/pyarrow/_parquet.pyx [hunk 2]
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
...
+                               self.null_count,
+                               self.distinct_count,
+                               self.num_values,
+                               self.physical_type)
+
+    def Reset(self):
+        self.statistics.get().Reset()
+
+    def SetComparator(self):
+        self.statistics.get().SetComparator()
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: Are either the Reset method or SetComparator method useful in Python? If not, let's remove them

File: python/pyarrow/_parquet.pyx [hunk 3]
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
...
+    def EncodeMin(self):
+        return self.statistics.get().EncodeMin()
+
+    def EncodeMax(self):
+        return self.statistics.get().EncodeMax()
+
+    def Encode(self):
+        statistics = EncodedStatistics()
+        statistics.init(self.statistics.get().Encode())
+        return statistics
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: Can you use snake case for these functions? `encode_min` etc. If these functions are not useful there's no immediate need to expose them

File: python/pyarrow/_parquet.pyx [hunk 4]
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
...
+
+    property file_offset:
+
+        def __get__(self):
+            return self.metadata.file_offset()
+
+    property file_path:
+
+        def __get__(self):
+            return self.metadata.file_path()
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: Use `frombytes` here to cast to platform string (for Py3)

File: python/pyarrow/_parquet.pyx [hunk 5]
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
...
+
+    property num_values:
+
+        def __get__(self):
+            return self.metadata.num_values()
+
+    property path_in_schema:
+
+        def __get__(self):
+            return self.metadata.path_in_schema().get().ToDotString()
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: `frombytes` (though I guess there's no guarantee about UTF8-ness)

File: python/pyarrow/_parquet.pyx [hunk 6]
```diff
@@ -52,6 +317,11 @@ cdef class RowGroupMetaData:
         self.metadata = self.up_metadata.get()
         self.parent = parent
 
+    def ColumnChunk(self, int i):
```
Comment (wesm) [inline python/pyarrow/_parquet.pyx]: Maybe just `column`?

File: python/pyarrow/tests/test_parquet.py [hunk 7]
```diff
@@ -442,6 +442,7 @@ def test_parquet_metadata_api():
     # Row group
     rg_meta = meta.row_group(0)
     repr(rg_meta)
+    repr(rg_meta.ColumnChunk(0))
```
Comment (wesm) [inline python/pyarrow/tests/test_parquet.py]: Can you add some tests to show that you can obtain the expected column statistics from each column in a row group? Checking the min and max for a string (bytes) column, integer, and double should be good enough. The idea of ARROW-480 was to be able to know when you can skip reading row groups based on a comparison predicate (like a value range)

**Review by wesm — COMMENTED** [review]
Thanks for doing this. Left some comments and asked about test coverage for the statistics part of this

**Comment by Licht-T** [comment]
@wesm Thanks!
- Removed unnecessary class and methods.
- Using Cython `bint` as C++ boolean type.
    http://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#types
- Renamed camel to snake.
- Converting C++ `std::string` to Python `str` by `frombytes`.
- Changed `EncodeMin`/`EncodeMax` to `min`/`max` decoded by `FormatStatValue`.
- Fixed `RowGroupStatistics.__repr__`

**Comment by Licht-T** [comment]
Now, I will make new tests!

**Comment by wesm** [comment]
hi @Licht-T, will you have time to finish this PR?

**Comment by Licht-T** [comment]
@wesm Yes! I'll do this in the couple of days.

**Comment by wesm** [comment]
@Licht-T how's it looking on this? Out of the PRs you have open, any that I or @cpcloud can help with?

**Comment by Licht-T** [comment]
@wesm Sorry for my late reply. Finally added the tests.

File: python/pyarrow/tests/test_parquet.py [hunk 8]
```diff
@@ -469,14 +469,62 @@ def test_parquet_metadata_api():
...
+        ([1, 2, 2, None, 4], np.uint32, u'1', u'4', 1, 4),
+        ([1, 2, 2, None, 4], np.uint64, u'1', u'4', 1, 4),
+        ([-1, 2, 2, None, 4], np.int16, u'-1', u'4', 1, 4),
+        ([-1, 2, 2, None, 4], np.int32, u'-1', u'4', 1, 4),
+        ([-1, 2, 2, None, 4], np.int64, u'-1', u'4', 1, 4),
+        ([-1.1, 2.2, 2.3, None, 4.4], np.float32, u'-1.1', u'4.4', 1, 4),
+        ([-1.1, 2.2, 2.3, None, 4.4], np.float64, u'-1.1', u'4.4', 1, 4),
+        (
+            [u'', u'b', unichar(1000), None, u'aaa'],
+            str, u' ', unichar(1000).encode('utf-8') + ' ', 1, 4
```
Comment (xhochy) [inline python/pyarrow/tests/test_parquet.py]: This line fails with Python 3, you probably need to change it to `, unichar(1000).encode('utf-8') + b' '`.

**Comment by Licht-T** [comment]
@xhochy Thanks! Fixed!

**Review by wesm — APPROVED** [review]
+1, thanks @Licht-T!

**Pull request closed without merging**